// Remember, send state, not differences
package wire;

// An update from the client to the server, sent over UDP
message ClientQuickUpdate {
    required sint32 x = 1;
    required sint32 y = 2;
    required sint32 angle = 3;
    // required sint32 angularVelocity = 4;
    required float velocityX = 5;
    required float velocityY = 6;
    // required float accelerationX = 7;
    // required float accelerationY = 8;
}

// An update from the server to the clients, sent over UDP
message ServerQuickUpdate {
    message PlayerUpdate {
        required uint32 playerID = 1;
        required ClientQuickUpdate update = 2;
    }
    
    repeated PlayerUpdate updates = 1;
}

// message ServerPlayerIDHandshake {
// required uint32 playerID = 1;
// }

message Chunk {
    required sint32 x = 1;
    required sint32 y = 2;
    
    // Tiles are compressed with snappy
    optional bytes tiles = 3;
    
    // Metadata is put into a "snake" (hilbert curve) then compressed with snappy
    optional bytes metadata = 4;
}
message Object {
    
    required sint32 x = 1;
    required sint32 y = 2;
    required sint32 angle = 3;
}
message Map {
    repeated Chunk chunks = 1;
    repeated Object objects = 2;
}
